lib/puppetserver/ca/revoke_action.rb

```diff
@@ -0,0 +1,123 @@
+      # (or a custom object that implements summarize() and help().
+      def self.parser(parsed = {})
+        OptionParser.new do |o|
+          o.banner = BANNER
+          o.on('--help', 'Display this revoke specific help output') do |help|
+            parsed['help'] = true
+          end
+          o.on('--version', 'Output the version') do |v|
+            parsed['version'] = true
+          end
+        end
+      end
+
+      def initialize(logger)
+        @logger = logger
+      end
+
+      # take unparsed array of arguments, return parsed hash and optionally an error code
+      def parse(args)
+        results = { certname: [] }
+        unparsed = []
+        parser = self.class.parser(results)
+
+        begin
+          parser.order!(args) do |unknown|
+            results[:certname] << unknown
+          end
+        rescue OptionParser::ParseError => e
+          unparsed += e.args
+          unparsed << args.shift unless args.first =~ /^-{1,2}/
+          retry
+        end
+
+        exit_code = nil
+        unless unparsed.empty?
+          @logger.err 'Error:'
+          @logger.err 'Unknown arguments or flags:'
+          unparsed.each do |arg|
+            @logger.err "    #{arg}"
+          end
+
+          @logger.err ''
+          @logger.err parser.help
+          exit_code = 1
+        end
+
+        if results[:certname].length == 0
+          @logger.err 'A certname to revoke is required'
+          @logger.err parser.help
+          exit_code = 1
+        elsif results[:certname].length > 1
+          @logger.err 'Only one certname is allowed at a time'
+          @logger.err 'Received certnames:'
+          results[:certname].each do |certname|
+            @logger.err "    #{certname}"
+          end
+          exit_code = 1
+
+        else
+          results[:certname] = results[:certname].first
+        end
+
+        return results, exit_code
+      end
+
+      # take the parsed arg hash from #parse and perform this action
+      # print information to @logger and return a valid exit code
+      def run(args)
+        certname = args[:certname]
+        body = JSON.dump({ desired_state: 'revoked' })
+        headers = {
+          'User-Agent' => 'PuppetserverCaCli',
+          'Content-Type' => 'application/json',
+          'Accept' => 'application/json'
+        }
+        uri = URI('https://localhost:8140/puppet-ca/v1/certificate_status/' +
+                  certname +
+                  '?environment=production')
+        store = OpenSSL::X509::Store.new
+        store.add_file(File.expand_path('~/.puppetlabs/etc/puppet/ssl/certs/ca.pem'))
+        store.add_file(File.expand_path('~/.puppetlabs/etc/puppet/ssl/certs/localhost.pem'))
+        store.add_file(File.expand_path('~/.puppetlabs/etc/puppet/ssl/crl.pem'))
+
+        result = Net::HTTP.start(uri.host, uri.port, use_ssl: true, cert_store: store) do |connection|
+          request = Net::HTTP::Put.new(uri, headers)
+          request.body = body
```

Comment:
There's not a lot of custom logic with this request so I'm hesitant to write much in the way of tests around it and leave the does-it-actually-talk-to-a-master test to acceptance level. I assume whatever helper we pull this logic into (eg an http util method or a CA class) we can stub it out in higher level specs.

Reply:
I think acceptance is fine here (though I think we should write a standalone test like we did for the [setup command](https://github.com/puppetlabs/puppetserver/blob/master/acceptance/suites/tests/certificate_authority/intermediate_ca.rb)). Because of how simple this is, I feel like introducing webmock would most just end up testing the HTTP library, which feels unnecessary.